File: Go/Solution with StringBuilder.go

package main

import (
    "fmt"
    "strings"
)

func compressedString(word string) string {
    compressed := strings.Builder{}
    frequency := 1

    for i := 1; i < len(word); i++ {
        if frequency == 9 || word[i-1] != word[i] {
            compressed.WriteByte(byte(frequency)+'0')
            compressed.WriteByte(word[i-1])
            frequency = 1
            continue
        }
        frequency++
    }
    compressed.WriteByte(byte(frequency)+'0')
    compressed.WriteByte(word[len(word)-1])

    return compressed.String()
}

File: C#/Solution with StringBuilder.cs

using System;

public class Solution
{
    public string CompressedString(string word)
    {
        var compressed = new StringBuilder(2 * word.Length);
        int frequency = 1;

        for (int i = 1; i < word.Length; ++i)
        {
            if (frequency == 9 || word[i - 1] != word[i])
            {
                compressed.Append((char)(frequency + '0'));
                compressed.Append(word[i - 1]);
                frequency = 1;
                continue;
            }
            ++frequency;
        }
        compressed.Append((char)(frequency + '0'));
        compressed.Append(word[word.Length - 1]);

        return compressed.ToString();
    }
}
